src/main/java/com/github/wojtechm/zadania_rekrutacyjne/string/ParenthesesCombinations.java

```diff
@@ -18,7 +20,58 @@
 @Difficulty(Level.HARD)
 class ParenthesesCombinations {
 
+    private ParenthesesCombinations(){}
+
     static List<String> parenthesesCombinations(int pairs) {
-        return null;
+        if (pairs <= 0){
+            return Collections.emptyList();
+        }
+
+        ArrayList<String> out = new ArrayList<>();
+        int number = 0;
+        for (int i = pairs; i > 0; i--) {
+            number += 1 << i * 2 - 1;
+        }
+
+        int maxNumber = (int) Math.pow(2, pairs * 2.0) - 1;
+        int mask = 0;
+        for (int i = 0; i < pairs; i++) {
+            mask += (int) Math.pow(2, i);
+        }
+        maxNumber &= mask << pairs;
+        int maxBeforeNumber = (int) (maxNumber - Math.pow(2, pairs - 1));
+
+        makeOut(pairs, out, number, maxBeforeNumber);
+        out.add(convert(maxNumber));
+        Collections.sort(out);
+        return out;
+    }
+
+    private static void makeOut(int pairs, ArrayList<String> out, int number, int maxBeforeNumber) {
+        while (number <= maxBeforeNumber) {
+            if (Integer.bitCount(number) == pairs) {
+                int sum = 0;
+                String converted = convert(number);
+                for (char item : converted.toCharArray()) {
+                    sum += (item == '(')? 1: -1;
+                    if (sum < 0) {
+                        break;
+                    }
+                }
+                if (sum == 0) {
+                    out.add(converted);
+                }
+            }
+            number += 2;
+        }
+    }
+
+    private static String convert(int in) {
```

Comment:
convert? Convert what? and to what? what is going on here?

Reply:
Done